modules/apps/frontend-js/frontend-js-spa-web/src/main/resources/META-INF/resources/senna/utils/pathParser.js

```diff
@@ -0,0 +1,184 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+const REGEX = /([/])?(?:(?::(\w+)(?:\(((?:\\.|[^\\()])*)\))?|\(((?:\\.|[^\\()])+)\))([+*?])?)/g;
+
+/**
+ * Converts the given array of regex matches to a more readable object format.
+ * @param {!Array<string>} matches
+ * @return {!Object}
+ */
+function convertMatchesToObj(matches) {
+	return {
+		match: matches[0],
+		modifier: matches[5],
+		name: matches[2],
+		paramPattern: matches[3],
+		prefix: matches[1],
+		unnamedPattern: matches[4],
+	};
+}
+
+/**
+ * Converts the given tokens parsed from a route format string to a regex.
+ * @param {!Array<string|!Object>} tokens
+ * @return {!RegExp}
+ */
+function convertTokensToRegex(tokens) {
+	let regex = '';
+	for (let i = 0; i < tokens.length; i++) {
+		if (typeof tokens[i] === 'string') {
+			regex += escape(tokens[i]);
+		}
+		else {
+			let capture = encloseNonCapturingGroup(tokens[i].pattern);
+			if (tokens[i].repeat) {
+				capture += encloseNonCapturingGroup('\\/' + capture) + '*';
+			}
+			capture = escape(tokens[i].prefix) + `(${capture})`;
+			if (tokens[i].optional) {
+				if (!tokens[i].partial) {
+					capture = encloseNonCapturingGroup(capture);
+				}
+				capture += '?';
+			}
+			regex += capture;
+		}
+	}
+
+	return new RegExp('^' + makeTrailingSlashOptional(regex) + '$');
+}
+
+/**
+ * Encloses the given regex pattern into a non capturing group.
+ * @param {string} pattern
+ * @return {string}
+ */
+function encloseNonCapturingGroup(pattern) {
+	return `(?:${pattern})`;
+}
+
+/**
+ * Escapes the given string to show up in the path regex.
+ * @param {string} str
+ * @return {string}
+ */
+function escape(str) {
+	return str.replace(/([.+*?=^!:${}()[]|\/\\])/g, '\\$1');
+}
+
+/**
+ * Makes trailing slash optional on paths.
+ * @param {string} regex
+ * @param {string}
+ */
+function makeTrailingSlashOptional(regex) {
+	if (/\/$/.test(regex)) {
```

Comment:
Would probably be more readable as `regex.endsWith('/')`.

Reply:
https://github.com/liferay-frontend/liferay-portal/pull/489#issuecomment-725325366